### ISSUE TITLE
Rename GridMap to MeshMap or TileMap3D and add actual GridMap node

### ISSUE THREAD

**Issue by Shadowblitz16** [godotengine/godot-proposals#789]
**Describe the project you are working on:**
Tower Defense

**Describe the problem or limitation you are having in your project:**
I currently have to write a custom grid and tile class for saving and loading due to there not being a actual grid map for nodes. 

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
Rename Gridmap to MeshMap or Tilemap3D and add actual Gridmap node for both 2d and 3d games.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
basicly it current Gridmaps would be renamed to something like Meshmap to be more clear and a true GridMap and GridMap2D would be added

the new Gridmaps would basically be Tilemaps or Meshmaps but would accept any node structure as long as the root was a Node2D for GridMap2D or Node3D for GridMap

this would allow people to still use old TileMaps or MeshMaps is they need performance

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
This would be used quite often and while it could be worked around its not something that should be.

**Is there a reason why this should be core and not an add-on in the asset library?:**
it would basicly be a better version of the Tilemap and current Gridmap but not limited to only sprites and meshes.

https://github.com/godotengine/godot-proposals/issues/747

**Comment by Calinou** [comment]
I think it makes sense to rename TileMap to TileMap2D and GridMap to TileMap3D.

**Comment by bojidar-bg** [comment]
This has come up multiple times in the past, both on [IRC](http://godot.eska.me/irc-logs/devel/2019-07-22.log) and on [GitHub](https://github.com/godotengine/godot/issues/30736#issuecomment-589997938).

Reduz mentioned that "problem is that tile is a very strict 2D concept" -- and that MeshLibrary could also be used "for terrain painting if we ever add that too" (can't link to logs).

I still feel it makes sense to unify those two, given that many people already refer to tileable 3D meshes as "tiles" (see https://www.kenney.nl/assets/tower-defense-kit and https://twitter.com/dwtw/status/810166761270243328) and that terrain painting might want to use a more complicated format (which specifies, say, permissible scales, custom instance variables/colors, etc.).

There is also a comment from Akien in the above-linked IRC log:
```
<Akien> Someone proposed TileMap2D and to rename GridMap to TileMap3D, but that doesn't make sense, they serve similar use cases but the APIs are different.
```

And indeed, if GridMap and TileMap are unified, their APIs would likely have to be unified too. With the plans to rework the tileset/tilemap to be more user-friendly and usable, unifying them would imply giving similar treatment to both.

---

This proposal has one more element: adding an "actual GridMap" which would allow efficiently placing 3D nodes on a grid. Some have argued that "Snap to Grid" already covers that, others have proposed allowing both meshes and non-meshes in the current GridMap node. For the moment, I will just link some discussions I can find.

* #747 - a proposal for adding functionality for placing arbitrary instances to GridMap.
    * https://github.com/godotengine/godot/issues/24096 - an issue discussing the same thing before the proposals repo was created.
* #493 a proposal for adding functionality for placing instances to TileMap (2D).
    * https://github.com/godotengine/godot/issues/9180 - an issue by me detailing the same. Note that this one has by far the most discussion.
    * https://github.com/godotengine/godot/issues/9180#issuecomment-442196887 - a nice comment by willnationsdev outlining an idea to have a GridPlacement resource which dictates snapping and thus can be reused between nodes.

**Comment by Shadowblitz16** [comment]
if you do add a grid map PLEASE allows us to specify a size for it.
infinite tile maps and grid maps are good in certain cases but sometimes we want to limit where things can be placed

**Comment by Calinou** [comment]
> infinite tile maps and grid maps are good in certain cases but sometimes we want to limit where things can be placed

Are you exposing editing tools to players? In this case, you're most likely writing your own editor where you can apply any restrictions you wish.

**Comment by Shadowblitz16** [comment]
@Calinou yes it was going to be a 3d ship builder. but the issue is that to limit to a specifies area I need to know the starting point of the grid (0,0) which is not easy with infinite grid maps.

also its a pretty basic feature to set the size of a grid. your basically already doing it internally.

**Comment by Mickeon** [comment]
What's the consensus on this PR, 2 years later?

**Comment by Calinou** [comment]
> What's the consensus on this PR, 2 years later?

Since feature freeze is ongoing, GridMap will be kept as-is for 4.0. Any significant changes will have to be made for 4.1, where *some* compatibility breakage will be tolerated as long as an upgrade path is provided.

**Comment by Meorge** [comment]
I wrote a proposal for effectively the same thing as is being described here, so rather than open a new issue in the repo for it, I'll just include what I had written about the topic here:

---

Godot offers the [TileMap node (current version named TileMapLayer for compatibility reasons)](https://docs.godotengine.org/en/stable/classes/class_tilemaplayer.html) to allow users to place 2D tiles and scenes on a grid. The 3D equivalent of this node is [GridMap](https://docs.godotengine.org/en/stable/classes/class_gridmap.html), allowing users to place 3D meshes on a grid instead.

TileMap's and GridMap's names do not convey their differences (namely, that they are for 2D and 3D respectively). For both of them, _tiles_ are placed on a _grid_. The documentation for both classes uses these terms to refer to them:

>  **TileMapLayer:** Node for 2D tile-based maps.
>  ...A TileMapLayer uses a TileSet which contain a list of tiles which are used to create grid-based maps.

> **GridMap:** Node for 3D tile-based maps.
> ...GridMap lets you place meshes on a grid interactively.

There was a similar situation with nodes in Godot 3. 2D nodes had the suffix "2D" (Node2D, Light2D, KinematicBody2D). 3D nodes did not have a suffix (Spatial, Light, KinematicBody). In Godot 4, the names were standardized, and the suffix was used to disambiguate between 2D and 3D counterparts (Node2D/3D, Light2D/3D, CharacterBody2D/3D).


At a point where large naming changes are acceptable (likely Godot 5.0), I think the TileMap/TileMapLayer and GridMap nodes should be renamed to be consistent with each other, differing only in their suffix. Possible naming pairs could include TileMap2D and TileMap3D or GridMap2D and GridMap3D.
